### PR TITLE
tasks 160- 172, without 161,163,170

### PR DESCRIPTION
**Pull request by IvanMenov**

### CODE REVIEW

File: src/main/java/com_github_leetcode/ListNode.java [hunk 1]
```diff
@@ -16,6 +16,12 @@ public ListNode(int val, ListNode next) {
         this.next = next;
     }
 
+    @Override
+    public boolean equals(Object obj) {
+        return this.val == ((ListNode) obj).val;
```
Comment (javadev) [inline src/main/java/com_github_leetcode/ListNode.java]: We compare ListNode by pointers not by value. The custom comparator may be applied for a solution.

File: src/main/java/com_github_leetcode/ListNode.java [hunk 2]
```diff
@@ -16,6 +18,25 @@ public ListNode(int val, ListNode next) {
...
+    public boolean equals(Object obj) {
+        if (this == obj) {
+            return true;
+        }
+        if (obj == null) {
+            return false;
+        }
+        if (getClass() != obj.getClass()) return false;
+        ListNode other = (ListNode) obj;
+        return val == other.val;
```
Comment (javadev) [inline src/main/java/com_github_leetcode/ListNode.java]: We need to compare ListNode by pointer, not by value.

**Comment by javadev** [comment]
There is an instruction for the readme.md generation tool:

https://github.com/javadev/LeetCode-in-Java/releases/tag/generatemd

The jar file itself:

https://github.com/javadev/LeetCode-in-Java/releases/download/generatemd/generatemd-1.0.jar

**Comment by IvanMenov** [comment]
Those codecov/patch, codecov/project check that were not successful was beacause the test cover only one scenario, as we agreed.

File: src/main/java/g0101_0200/s0160_intersection_of_two_linked_lists/Solution.java [hunk 3]
```diff
@@ -7,7 +7,7 @@ public ListNode getIntersectionNode(ListNode headA, ListNode headB) {
         ListNode pA = headA;
         ListNode pB = headB;
         while (pA != null || pB != null) {
-            if (pA != null && pB != null && pA.equals(pB)) {
+            if (pA != null && pB != null && pA.toString().equals(pB.toString())) {
```
Comment (javadev) [inline src/main/java/g0101_0200/s0160_intersection_of_two_linked_lists/Solution.java]: This solution works for me:

```java
public class Solution {
    public ListNode getIntersectionNode(ListNode headA, ListNode headB) {
        ListNode node1 = headA;
        ListNode node2 = headB;
        while (node1 != node2) {
            node1 = node1 == null ? headB : node1.next;
            node2 = node2 == null ? headA : node2.next;
        }
        return node1;
    }
}
```

File: src/main/java/g0101_0200/s0160_intersection_of_two_linked_lists/Solution.java [hunk 4]
```diff
@@ -4,15 +4,12 @@
...
-        ListNode pB = headB;
-        while (pA != null || pB != null) {
-            if (pA != null && pB != null && pA.toString().equals(pB.toString())) {
-                return pA;
-            }
-            pA = pA == null ? headB : pA.next;
-            pB = pB == null ? headA : pB.next;
+        ListNode node1 = headA;
+        ListNode node2 = headB;
+        while (!node1.toString().equals(node2.toString())) {
```
Comment (javadev) [inline src/main/java/g0101_0200/s0160_intersection_of_two_linked_lists/Solution.java]: Why toString?

Comment (IvanMenov) [reply src/main/java/g0101_0200/s0160_intersection_of_two_linked_lists/Solution.java]: The above implementation, that you wrote results in nullpoinerexception in the test case, as node1 == node2 evaluates to false

Comment (javadev) [reply src/main/java/g0101_0200/s0160_intersection_of_two_linked_lists/Solution.java]: I tested it in leetcode. There are no exceptions.

image

File: src/main/java/g0101_0200/s0172_factorial_trailing_zeroes/Solution.java [hunk 5]
```diff
@@ -0,0 +1,15 @@
+package g0101_0200.s0172_factorial_trailing_zeroes;
+
+public class Solution {
+    private Solution() {}
+
+    public static int trailingZeroes(int n) {
```
Comment (javadev) [inline src/main/java/g0101_0200/s0172_factorial_trailing_zeroes/Solution.java]: Why static?

File: src/test/java/g0101_0200/s0160_intersection_of_two_linked_lists/SolutionTest.java [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+    public void getIntersectionNode() {
+        ListNode nodeA =
+                new ListNode(4, new ListNode(1, new ListNode(8, new ListNode(4, new ListNode(5)))));
+        ListNode nodeB =
+                new ListNode(
+                        5,
+                        new ListNode(
+                                6,
+                                new ListNode(
+                                        1, new ListNode(8, new ListNode(4, new ListNode(5))))));
```
Comment (javadev) [inline src/test/java/g0101_0200/s0160_intersection_of_two_linked_lists/SolutionTest.java]: It should be the same ListNode(8) for both lists.

File: src/main/java/g0101_0200/s0165_compare_version_numbers/Solution.java [hunk 7]
```diff
@@ -0,0 +1,45 @@
+package g0101_0200.s0165_compare_version_numbers;
+
+public class Solution {
+
+    public int compareVersion(String version1, String version2) {
+        String[] ver1 = version1.split("\\.", 0);
+        String[] ver2 = version2.split("\\.", 0);
```
Comment (javadev) [inline src/main/java/g0101_0200/s0165_compare_version_numbers/Solution.java]: It is better to use solution without split.

```java
class Solution {
    public int compareVersion(String version1, String version2) {
        // acquire first number
        int numA = 0;
        int i;
        for(i=0; i<version1.length(); i++){
            char c = version1.charAt(i);
            if(c == '.'){
                break;
            }else{
                numA = numA*10 + ((int)c - 48);
            }
        }
        
        // acquire second number
        int numB = 0;
        int j;
        for(j=0; j<version2.length(); j++){
            char c = version2.charAt(j);
            if(c == '.'){
                break;
            }else{
                numB = numB*10 + ((int)c - 48);
            }
        }
        
        // compare
        if(numA > numB){
            return 1;
        }else if(numA < numB){
            return -1;
        }else{ // equal
            String v1 = "";
            String v2 = "";

            if(i != version1.length())
                v1 = version1.substring(i+1);

            if(j != version2.length())
                v2 = version2.substring(j+1);
            
            // if both versions end here, they are equal
            if(v1.equals("") && v2.equals("")){
                return 0;
            }else{
                return compareVersion(v1, v2);
            }
        }
    }
}
```

**Pull request merged**